near/src/runtime.rs

```diff
@@ -811,17 +816,24 @@ impl RuntimeAdapter for NightshadeRuntime {
             cursor.read_exact(&mut value)?;
             state_update.set(key, DBValue::from_slice(&value));
         }
-        let (store_update, root) = state_update.finalize()?.into(self.trie.clone())?;
-        if root != state_root.hash {
-            return Err("Invalid state root".into());
-        }
+        let (store_update, _) = state_update.finalize()?.into(self.trie.clone())?;
         store_update.commit()?;
         Ok(())
     }
 
-    fn confirm_state(&self, _state_root: &StateRoot) -> Result<bool, Error> {
-        // TODO approve that all parts are here
-        Ok(true)
+    fn confirm_state(&self, state_root: &StateRoot) -> Result<(), Box<dyn std::error::Error>> {
+        // TODO Misha: compare self.trie.root() and state_root.hash directly if possible
+        let mut state_update = TrieUpdate::new(self.trie.clone(), CryptoHash::default());
+        for item in self.trie.iter(&state_root.hash)? {
+            let (key, value) = item?;
+            state_update.set(key, DBValue::from_slice(&value));
+        }
+        let (_, root) = state_update.finalize()?.into(self.trie.clone())?;
+        // It's enough to ensure that state_root.hash is equal to trie root to prove that all parts are accepted
+        if root != state_root.hash {
+            return Err("Invalid state root".into());
+        }
```

Comment:
this doesn't need to exist, parts validation should take care of everything.

should still have a method somewhere that removes existing trie and replaces it with the one built from parts.